lib/Core/BuildEngine.cpp

```diff
@@ -1359,10 +1359,12 @@ class BuildEngineImpl : public BuildDBDelegate {
   RuleInfo& addRule(KeyID keyID, Rule&& rule) {
     auto result = ruleInfos.emplace(keyID, RuleInfo(keyID, std::move(rule)));
     if (!result.second) {
-      // FIXME: Error handling.
-      std::cerr << "error: attempt to register duplicate rule \""
-                << rule.key << "\"\n";
-      exit(1);
+      delegate.error("error: attempt to register duplicate rule \"" + rule.key + "\"\n");
```

Comment:
I think the `error: ` prefix here is not needed since the delegate knows it's an error when receiving the method call. It should then decide on its own how to present this error.

Reply:
I had dropped it, though all of the error messages below in the source file sent to the delegate also had the prefix, so I put it back.